### ISSUE TITLE
Async Actions Race Condition

### ISSUE THREAD

**Issue by bolein** [marcglasberg/async_redux#30]
What is the execution order of the async reducers (actions) dispatched simultaneously? Is it sequential or parallel?

**Comment by marcglasberg** [comment]
A reducer is only sync if both the `reducer` AND the `before` methods are sync. If you make any of them async, then the reducer is async. Also note, you make the reducer sync or async by returning `AppState` or `Future<AppState>`, respectively.

1) When you dispatch sync reducers, they are executed synchronously, in the order they are called.  For example, this code will wait until `MyAction1` is finished and only then run `MyAction2` (assuming both are sync reducers): 

    dispatch(MyAction1()); 
    dispatch(MyAction2());

2) When you dispatch an async reducer with `dispatch` it starts immediately, but there is really no way of knowing when it finishes, because it depends on how long the async code takes to run, which depends on the async code itself, not AsyncRedux. 

Dispatching an async reducer with `dispatch` is like starting any other async processes without writing `await`. The process will start, but you are not waiting for it to finish. For example, this code will start both `MyAsyncAction1` and `MyAsyncAction2`, but there is no way of knowing how long they will take or which one finishes first:

    dispatch(MyAsyncAction1()); 
    dispatch(MyAsyncAction2());

3) If you want to wait for some async action to finish, you must dispatch it with `dispatchFuture` instead of `dispatch`. Then you can actually wait for it to finish. For example, this code will **wait** until `MyAsyncAction1` is finished, and **only then** run `MyAsyncAction2`: 

    await dispatchFuture(MyAsyncAction1()); 
    await dispatchFuture(MyAsyncAction2())`.

**Comment by bolein** [comment]
Thanks for the detailed response

**Comment by cgestes** [comment]
(not 100% percent sure it belongs here, but this made me tick..)
If my understanding is correct it means that you should not return state in async reducers to avoid races. Only sync reducers should be used to change the state. Async reducers should dispatch another action to modify the state, because other actions can execute between the return of the reducer and the actual application of the returned state on the store. State that could have been modified by another dispatched action in the meantime. This would mean the changes made to the state in the meantimes would be lost.

**Comment by marcglasberg** [comment]
@cgestes That's **not** right. Async reducers dispatching another action to modify the state yields the exact same result as the Async reducer returning itself the new state. There is absolutely no need to dispatch another action to modify the state. Changes made to the state are never lost.

When you return the state it will get the most recent state in the store, not a copy of the state from when the async reducer started. In other words, using `state` is actually a getter from `store.state`, and will thus always contain the most recent state.

Maybe I should make this more clear in the documentation.

**Comment by bolein** [comment]
I think @cgestes is actually right. Even using the getter async reducers become indeterministic because the state may change at any point of time throughout the execution. 

It would make sense to use an implicit contract that state mutations should only happen in sync reducers (because they are executed sequentially and the state won't change externally during the execution). But this basically returns us to the idea of middleware.

**Comment by cgestes** [comment]
let me try to explain the race again:

You have the following function in async_redux:
```dart
FutureOr<void> _applyReducer(ReduxAction<St> action) {
    _reduceCount++;

    var result = action.reduce();

    if (result is Future<St>) {
      return result.then((state) => _registerState(state, action));
    } else if (result is St) {
      St state = result;
      _registerState(state, action);
    } else if (result != null) {
      throw AssertionError();
    }
  }
```

The following code is problematic:
```dart
    var result = action.reduce();
    // anything can happen here while we wait for the future to complete, it can be other dispatch that affect the state
    // everything that have been changed waiting for the future to complete will be erased by the following call to _registerState, cause the former action.reduce() have been run on an old state.
    if (result is Future<St>) {
      return result.then((state) => _registerState(state, action));
```

**Comment by marcglasberg** [comment]
Sorry, but that's not the case, and you are both misinterpreting what async reducers do or were created to do. I guarantee you there is no way state may be lost, and async reducers are also deterministic insofar as async code in general can be deterministic. 

Instead of discussing with words we should be looking at code. I will close this issue and open another one about documenting how state is changed in async reducers, and creating code to test/demonstrate those. Then, by looking at simple examples, and being able to run them, this discussion is going to be easier, and you both will be able to modify the examples to make any point you want. Is that ok?

**Comment by marcglasberg** [comment]
Done: https://github.com/marcglasberg/async_redux/issues/31 

As soon as I create this examples I will let you both know and we can restart from there.

**Comment by cgestes** [comment]
I have a feeling that you believe we are stupid.

So let me answer following the same style you did in that thread:

> Sorry, but that's not the case, and you are both misinterpreting what async reducers do or were created to do.
You are wrong (tm), there is a race, and by the way what they were created for is irrelevant here.

>  I guarantee you there is no way state may be lost, and async reducers are also deterministic insofar as async code in general can be deterministic.
It depends on the dart VM and the implementation of async / await. Async code is generally underterministic. BTW I proved you above that states can be lost, I also experimented it locally.

Anyway good luck. I fixed my race and do not depends on async_redux anymore.
